### PR TITLE
To make save/load compatible between dygraph and static-graph.

### PR DESCRIPTION
**Pull request by guoshengCS**
To make save/load compatible between dygraph and static-graph.

由于静态图和动态图save内容存在以下不一致，在load时进行转换处理：

- global step：使用learning rate schduler时
  - 动态图保存的key为"global_step" to save, 静态图保存的key为"@LR_DECAY_COUNTER@".
  - 由于global step自增时机不同，动态图在最后，静态图在最前（初始-1），动态图保存的值会比动态图多1

- global lr：未使用learning rate schduler时
  - 静态图中作为persistable var保存，动态图不保存

- accumulator name：`unique(param.name + "_" + optimizer.name+ "_" + accum_name)`
  - 静态图和动态图中optimizer.name不一致
https://github.com/PaddlePaddle/Paddle/blob/release/1.7/python/paddle/fluid/optimizer.py#L69

此外验证过程发现以下待修复问题：

- [x] 静态图eval之后的train结果错误，原因是由于learning rate schedule在ori_prog中定义，eval_prog同样clone了下来 https://github.com/PaddlePaddle/hapi/blob/master/model.py#L291
- [x] 静态图对input使用重载运算错误，原因是由于重载运算使用var的Program，但使用的var为ori_prog中的var
- [x] 静态图[load到CPU]( https://github.com/PaddlePaddle/hapi/blob/master/model.py#L200)，只有第一步计算结果正确， 原因是由于sgd等optimizer类OP的输入输出inplace和transfer一起使用时的bug，transfer back会覆盖sgd的输出。https://github.com/PaddlePaddle/Paddle/pull/22936

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -197,7 +198,10 @@ def _load(path):
         assert param_state, "failed to load parameters, please check path"
 
         if self._executor is None:
-            executor = fluid.Executor(fluid.CPUPlace())._default_executor
+            # TODO: loading to CPU seems to some transform error, and only
```
Comment (willthefrog) [inline model.py]: how does it behave exactly? Is it an regression in the framework code? I think it works earlier, but not completely sure.

File: model.py [hunk 2]
```diff
@@ -289,6 +346,14 @@ def _run(self, inputs, labels=None, device='CPU', device_ids=None):
...
     def _make_program(self, inputs):
         prog = self._orig_prog.clone()
+        # change inputs to the same var in cloned program
+        inputs = fluid.layers.utils.map_structure(
+            lambda var: prog.global_block().var(var.name), inputs)
+        # prune unraleted ops in test program, mainly for ops inserted
+        # by learning rate scheduling
+        if self.mode != 'train':
+            for op in list(prog.global_block().ops):
+                prog.global_block()._remove_op(0)
```
Comment (willthefrog) [inline model.py]: quite hard to get one's head around this part, need more clarification in comments, If I am understanding it correctly:
- the learning rate  schedulers insert vars when constructed, which happens before first run
- this line remove all ops in the program, since it is before `forward`, the program contains only optimizer related vars
- one thing I am not so sure about: if the LR scheduler vars were not to be removed, how will it affect the training process? (the counter will increment even without`minimize` call?)

File: model.py (hunk 1 above)
Comment (guoshengCS) [reply model.py]: Run the appendix code as following:

 1. save and train (using dygraph to save before training)

```
[array([1.4861668], dtype=float32)]
[array([1.659631], dtype=float32)]
```

2. load with the original cpu executor and train （static-graph）

```
[array([1.4861668], dtype=float32)]
[array([1.6598837], dtype=float32)]
```

3. load with the gpu executor and train （static-graph）

```
[array([1.4861668], dtype=float32)]
[array([1.659631], dtype=float32)]
```

When loading with the cpu executor,  only the first step gets a right result (the results should be same with that under 1).  It also occurs on Transformer.


Appendix code:

```python
from __future__ import division
from __future__ import print_function

import contextlib
import os

import numpy as np

import paddle
from paddle import fluid
from paddle.fluid.dygraph.nn import Linear

from model import Model, CrossEntropy, shape_hints

random_seed = 123
np.random.seed(random_seed)
batch_size = 8
in_size = 32
class_num = 4

@contextlib.contextmanager
def null_guard():
    yield

guard = fluid.dygraph.guard() if True else null_guard()

class MyModel(Model):
    def __init__(self):
        super(MyModel, self).__init__()
        self._fc = Linear(in_size, class_num, act="softmax")

    def forward(self, x):
        return self._fc(x)

with guard:
    fluid.default_main_program().random_seed = random_seed
    fluid.default_startup_program().random_seed = random_seed

    model = MyModel()
    optim = fluid.optimizer.SGD(0.001, parameter_list=model.parameters())
    model.prepare(optim, CrossEntropy())
    #model.save("linear_dygraph_tmp/model")
    #model.load("linear_dygraph_tmp/model")

    for i in range(2):
        input = np.random.rand(batch_size, in_size).astype("float32")
        target = np.random.randint(0, class_num, (batch_size, 1)).astype("int64")
        out, loss = model.train(input, target, device='gpu', device_ids=[0])
        print(loss)
```

File: model.py (hunk 2 above)
Comment (guoshengCS) [reply model.py]: Done.  Update the comments. 

> if the LR scheduler vars were not to be removed, how will it affect the training process? (the counter will increment even withoutminimize call?)

As you said, the counter will increment even without minimize call . And we can `print(fluid.default_main_program())`  to inspect the ops to increase global step var after `model.prepare`.

File: model.py (hunk 1 above)
Comment (guoshengCS) [reply model.py]: This is caused by transfer for inplace, however it does not run as expected. 
Fix it by https://github.com/PaddlePaddle/Paddle/pull/22936

File: model.py (hunk 2 above)
Comment (willthefrog) [reply model.py]: okay, I see

File: model.py [hunk 3]
```diff
@@ -197,7 +198,10 @@ def _load(path):
         assert param_state, "failed to load parameters, please check path"
 
         if self._executor is None:
-            executor = fluid.Executor(fluid.CPUPlace())._default_executor
+            # TODO: loading to CPU seems to some transform error, and only
+            # the first step get the right result
+            # executor = fluid.Executor(fluid.CPUPlace())._default_executor
+            executor = fluid.Executor(fluid.CUDAPlace(0))._default_executor
```
Comment (xyzhou-puck) [inline model.py]: why cudaplace?

File: model.py [hunk 4]
```diff
@@ -289,6 +346,17 @@ def _run(self, inputs, labels=None, device='CPU', device_ids=None):
 
     def _make_program(self, inputs):
         prog = self._orig_prog.clone()
+        # change inputs to the same var in cloned program
+        inputs = fluid.layers.utils.map_structure(
+            lambda var: prog.global_block().var(var.name), inputs)
+        # NOTE: When defining learning rate scheduling in static-graph, ops to
```
Comment (willthefrog) [inline model.py]: just put it out there, maybe the correct approach is to fix it in the framework.

to summarize, this is a workaround for some bug/non-intuitive behavior in paddle:
1.  LR scheduler would insert auto increment counter op in global scope when constructed, however the op does not have a `backward` or `optimizer` role, so it runs every step even without `optimizer.minimize` call
2. `program.clone(for_test=True)` gives the impression that it would generated a test program from a training program, but in reality it simple toggles test state for some ops (e.g., bn, dropout), and does not prune any optimizer related stuff. In practice, instead of clone an existing training program, user would have to construct a test program from scratch and then clone it for test.

File: model.py (hunk 3 above)
Comment (guoshengCS) [reply model.py]: Done. Soved by https://github.com/PaddlePaddle/Paddle/pull/22936
